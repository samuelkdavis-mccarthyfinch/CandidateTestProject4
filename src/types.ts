import { termSymbol } from './constants';

export interface Term {
  value: string;
  [termSymbol]: true;
}

export interface ConceptNode {
  '@id': string;
  '@type': NodeType.Node;
  label: string;
  language: Language;
  sense_label: string;
  term: string;
}

export interface Relation {
  '@id': RelationId;
  '@type': NodeType.Relation;
  label: string;
}

export interface Edge {
  '@id': string;
  '@type': NodeType.Edge;
  dataset: string;
  end: ConceptNode;
  license: string;
  rel: Relation;
  sources: Array<Object>;
  start: ConceptNode;
  surfaceText: null;
  weight: number;
}

export enum RelationId {
  IsA = '/r/IsA',
}

export enum Language {
  English = 'en',
}

export enum NodeType {
  Relation = 'Relation',
  Edge = 'Edge',
  Node = 'Node',
}

export interface TermRelationships {
  term: Term;
  ancestors: Record<string, boolean>;
}

export interface Relationships {
  [key: string]: Relationships;
}

export interface ConceptNetApi {
  getAncestralEdges: (term: Term) => Promise<Array<Edge>>;
  isParentOfATerm: (args: { parent: Term; child: Term }) => Promise<boolean>;
}
