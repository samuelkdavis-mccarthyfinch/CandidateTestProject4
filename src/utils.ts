import { termSymbol } from './constants';
import { Term, Language } from './types';

export const asTerm = (termString: string): Term => {
  return {
    value: termString,
    [termSymbol]: true,
  };
};

export const conceptToTerm = (concept: string, language: string = Language.English): Term => ({
  value: `/c/${language}/${concept.replace(/ /g, '_')}`,
  [termSymbol]: true,
});
