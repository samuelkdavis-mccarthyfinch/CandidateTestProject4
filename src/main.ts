import process from 'process';
import util from 'util';

import { getApi } from './api';
import { Term, TermRelationships, Relationships, ConceptNetApi } from './types';
import { asTerm, conceptToTerm } from './utils';

export const getTermRelationships =
  (deps: { api: ConceptNetApi } = { api: getApi() }) =>
  async (term: Term) => {
    const edges = await deps.api.getAncestralEdges(term);
    const termRelationships: Array<TermRelationships> = edges.map((x) => {
      return {
        term: asTerm(x.end.term),
        ancestors: {},
      };
    });

    for (const parentTerm of termRelationships) {
      for (const childTerm of termRelationships) {
        if (parentTerm === childTerm) {
          continue;
        }

        if (
          await deps.api.isParentOfATerm({
            parent: parentTerm.term,
            child: childTerm.term,
          })
        ) {
          childTerm.ancestors[parentTerm.term.value] = true;
        }
      }
    }

    termRelationships.push({
      term: term,
      ancestors: edges.reduce(
        (sum: Record<string, boolean>, x) => ({
          ...sum,
          [x.end.term]: true,
        }),
        {} as Record<string, boolean>,
      ),
    });

    return termRelationships;
  };

export const buildHierarchy = (root: TermRelationships, termRelationships: Array<TermRelationships>): Relationships => {
  const child = termRelationships.find((x) => x.ancestors[root.term.value]);

  return {
    [root.term.value]: child ? buildHierarchy(child, termRelationships) : {},
  };
};

export async function findConceptHierarchiesForConcept(concept: string) {
  const term = conceptToTerm(concept);
  const termRelationships = await getTermRelationships()(term);
  const roots = termRelationships.filter((x) => Object.keys(x.ancestors).length === 0);

  if (roots.length === 0) {
    throw new Error('No root nodes were found');
  }

  const chains = roots.map((x) => buildHierarchy(x, termRelationships));
  return chains;
}

export async function main() {
  const [, , concept] = process.argv;
  if (!concept) {
    throw new Error('Please provide a concept you want to lookup as an argument to the program');
  }
  console.log(`Looking up concept hierarchy for ${concept}`);

  const chains = await findConceptHierarchiesForConcept(concept);
  console.log(util.inspect(chains, { depth: 15 }));
}
