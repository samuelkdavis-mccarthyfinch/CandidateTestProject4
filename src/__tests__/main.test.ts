import { TermRelationships } from '../types';
import { asTerm } from '../utils';
import { buildHierarchy, getTermRelationships } from '../main';

describe('buildHierarchy', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  test('correct result', async () => {
    const termRelationshipRoot: TermRelationships = { term: asTerm('term_1'), ancestors: {} };
    const termRelationship2: TermRelationships = {
      term: asTerm('term_2'),
      ancestors: { [termRelationshipRoot.term.value]: true },
    };

    const termRelationship3: TermRelationships = {
      term: asTerm('term_3'),
      ancestors: { [termRelationship2.term.value]: true },
    };

    const result = buildHierarchy(termRelationshipRoot, [termRelationshipRoot, termRelationship2, termRelationship3]);

    expect(result).toMatchSnapshot();
  });
});

describe('getTermRelationships', () => {
  const mockApi = {
    getAncestralEdges: jest.fn(),
    isParentOfATerm: jest.fn(),
  };

  beforeEach(() => {
    jest.resetAllMocks();
  });

  test('correct result', async () => {
    const edge1 = {
      end: {
        term: 'term_1',
      },
    };

    const edge2 = {
      end: {
        term: 'term_2',
      },
    };

    const edge3 = {
      end: {
        term: 'term_3',
      },
    };

    mockApi.getAncestralEdges.mockResolvedValueOnce([edge1, edge2, edge3]);

    mockApi.isParentOfATerm.mockImplementation((args) => {
      if (args.parent.value === edge1.end.term && args.child.value === edge2.end.term) {
        return true;
      }

      if (args.parent.value === edge2.end.term && args.child.value === edge3.end.term) {
        return true;
      }

      return false;
    });

    const term = asTerm('term');

    const result = await getTermRelationships({ api: mockApi })(term);

    expect(result).toMatchSnapshot();
  });
});
