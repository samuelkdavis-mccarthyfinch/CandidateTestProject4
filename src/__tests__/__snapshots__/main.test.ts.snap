// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildHierarchy correct result 1`] = `
Object {
  "term_1": Object {
    "term_2": Object {
      "term_3": Object {},
    },
  },
}
`;

exports[`getTermRelationships correct result 1`] = `
Array [
  Object {
    "ancestors": Object {},
    "term": Object {
      "value": "term_1",
      Symbol(term): true,
    },
  },
  Object {
    "ancestors": Object {
      "term_1": true,
    },
    "term": Object {
      "value": "term_2",
      Symbol(term): true,
    },
  },
  Object {
    "ancestors": Object {
      "term_2": true,
    },
    "term": Object {
      "value": "term_3",
      Symbol(term): true,
    },
  },
  Object {
    "ancestors": Object {
      "term_1": true,
      "term_2": true,
      "term_3": true,
    },
    "term": Object {
      "value": "term",
      Symbol(term): true,
    },
  },
]
`;
